core/src/main/java/org/infinispan/container/impl/L1DefaultInternalDataContainer.java

```diff
@@ -27,10 +27,10 @@
  * @author wburns
  * @since 9.3
  */
-public class L1DefaultSegmentedDataContainer<K, V> extends DefaultSegmentedDataContainer<K, V> {
+public class L1DefaultInternalDataContainer<K, V> extends DefaultInternalDataContainer<K, V> {
```

Comment:
Sounds good.

Reply:
Hrmm, actually looking at this, I can't call `clear(IntSet)` here. This would have to be done before rehash starts. Otherwise I would be removing now valid entries, possibly.

Reply:
And actually calling `clear(IntSet)` earlier means this will throw off the ExceptionEviction count I believe. I think I am going to just add it back as a listener and remove on any topology update.

Also I should put a thing to prevent a user from configuring L1 with ExceptionEviction as it doesn't make much sense.

Reply:
Hrmm, yeah keeping the `removeSegments` invocation and just remove the empty check wouldn't require the listener, so I can do that 👍

Reply:
+1 to disable exception eviction with L1, both failing a read because the local DC is full and failing a regular write to keep an L1 entry instead would be bad.

-1 to make it a listener again, a better option may be to merge `addSegments()` and `removeSegments()` into a `setOwnedSegments()` method and clear the L1 map there.

Reply:
I logged https://issues.jboss.org/browse/ISPN-9297 re: L1 and exception eviction.

Also talking with @danberindei, using the `removeSegments` approach seems best.